src/compiler/array.rs

```diff
@@ -224,3 +224,363 @@ pub(crate) fn get(compiler: &mut Compiler, arr: &Symbol, index: &Symbol) -> Symb
+                // [i, i]
+                Instruction::MemLoad(Some(length(arr).memory_addr)),
+                // [len, i, i]
+                Instruction::U32CheckedLT,
+                // [i < len, i]
+            ],
+            body: vec![
+                Instruction::Dup(None),
+                // [i, i]
+                Instruction::MemStore(Some(current_element_index.memory_addr)),
+                // [i]
+                Instruction::Dup(None),
+                // [i, i]
+                Instruction::Push(current_element.type_.miden_width()),
+                // [inner_width, i, i]
+                Instruction::U32CheckedMul,
+                // [offset = i * inner_width, i]
+                Instruction::MemLoad(Some(data_ptr(arr).memory_addr)),
+                // [data_ptr, offset, i]
+                Instruction::U32CheckedAdd,
+                // [ptr = data_ptr + offset, i]
+            ]
+            .into_iter()
+            .chain((0..current_element.type_.miden_width()).flat_map(|y| {
+                vec![
+                    Instruction::Dup(None),
+                    // [ptr, ptr, i]
+                    Instruction::Push(y),
+                    // [y, ptr, ptr, i]
+                    Instruction::U32CheckedAdd,
+                    // [ptr + y, ptr, i]
+                    Instruction::MemLoad(None),
+                    // [value, ptr, i]
+                    Instruction::MemStore(Some(current_element.memory_addr + y)),
+                    // [ptr, i]
+                ]
+            }))
+            .chain([Instruction::Drop])
+            .chain(body)
+            .chain([
+                Instruction::Push(1),
+                // [1, i]
+                Instruction::U32CheckedAdd,
+                // [i = i + 1]
+            ])
+            .collect(),
+        },
+        // [i]
+        Instruction::Drop,
+    ]);
+
+    Ok(())
+}
+
+fn grow(compiler: &mut Compiler, arr: &Symbol, needed_len: &Symbol) -> Result<Symbol> {
+    ensure_eq_type!(arr, Type::Array(_));
+    ensure_eq_type!(needed_len, Type::PrimitiveType(PrimitiveType::UInt32));
+
+    let result = compiler
+        .memory
+        .allocate_symbol(Type::PrimitiveType(PrimitiveType::Boolean));
+
+    let then_instructions = {
+        let mut insts = Vec::new();
+        std::mem::swap(compiler.instructions, &mut insts);
+
+        let new_capacity = compiler
+            .memory
+            .allocate_symbol(Type::PrimitiveType(PrimitiveType::UInt32));
+        compiler
+            .memory
+            .read(compiler.instructions, array::length(arr).memory_addr, 1);
+        // [len]
+        compiler.instructions.push(Instruction::Push(2));
+        compiler
+            .instructions
+            .push(encoder::Instruction::U32CheckedMul);
+        // [len * 2]
+        compiler.instructions.push(Instruction::Push(16));
+        compiler
+            .instructions
+            .push(encoder::Instruction::U32CheckedAdd);
+        // [len * 2 + 15]
```

Comment:
We have a lot of memory available in Miden VM (2^32), but reading/writing it is slow, so it's better to over-allocate to avoid having to allocate again to grow the array and copy over items. `16` is just how much I thought would be enough for a new empty array that will be `push`ed to, should be enough for most use-cases. Side note: I wonder if using linked lists would be more efficient.